### PR TITLE
Fix issue 19949: Implement support for C++ ABI tags

### PR DESCRIPTION
**Pull request by Geod24**
```
Implements C++ ABI tags with a few more restriction than the previous PR (9995).
In particular, do not follow C++ source-level conventions when impractical
(e.g. propagation). Since this is a feature that will be seldom used by users,
I favored implementation simplicity over user-friendly semantics.
```

Previous PR by @SSoulaimane : https://github.com/dlang/dmd/pull/9995
Since the author has gone AWOL and there were some discussions on the initial PR, I had a go at implementing it. I mostly used the tests from the original PR, which were very comprehensive, but tried  to keep the special casing in semantic to a minimum.
Requires https://github.com/dlang/druntime/pull/2990 to pass tests.

### CODE REVIEW

**Comment by thewilsonator** [comment]
`restruction`? Restriction?

File: src/dmd/semantic2.d [hunk 1]
```diff
@@ -596,3 +614,101 @@ private extern(C++) final class Semantic2Visitor : Visitor
...
+        // Valid element
+    }
+    // Since ABI tags need to be sorted, we sort them in place
+    // It might be surprising for users that inspects the UDAs,
+    // but it's a concession to practicality.
+    // Casts are unfortunately necessary as `implicitConvTo` is not
+    // `const` (and nor is `StringExp`, by extension).
+    static int predicate(const scope Expression* e1, const scope Expression* e2) nothrow
+    {
+        scope(failure) assert(0, "An exception was thrown");
```
Comment (thewilsonator) [inline src/dmd/semantic2.d]: dmd doesn't throw, so why this?

File: src/dmd/cppmangle.d [hunk 2]
```diff
@@ -935,16 +974,30 @@ private final class CppMangleVisitor : Visitor
...
+            buf.writestring("_Z");
+            source_name(d, false);
+        }
         else
         {
-            if (!isNested)
-                buf.writestring(d.ident.toString());
-            else
+            auto varTags = ABITagContainer.forSymbol(d);
+            if (varTags)
```
Comment (thewilsonator) [inline src/dmd/cppmangle.d]: why not just
```
if (ABITagContainer.forSymbol(d))
```

Comment (Geod24) [reply src/dmd/cppmangle.d]: Indeed, this is a leftover of a previous iteration, will fix.

File: src/dmd/semantic2.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/semantic2.d]: Might be that some of those functions are not annotated as `nothrow`, will double check and fix as needed.

**Comment by Geod24** [comment]
So on OSX, the compiler is too old:
```
HOST_CXX (c++): Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)
Target: x86_64-apple-darwin13.4.0
Thread model: posix
```

From a quick Google, only LLVM 3.9 [knows about ABI tags](https://releases.llvm.org/3.9.0/tools/clang/docs/ReleaseNotes.html#id4).
 3.9 was release 2016-09-02, 3.5.0 was released 2014-01-02 and 3.5.2 was released 2015-04-02.

I tested it on my machine, it works, but there's no way to test it on the auto-tester.

File: src/dmd/semantic2.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/semantic2.d]: Okay I realize why I didn't do this: It's a giant rabbit hole. Same issue with `const`. You'd have to make most of DMD `nothrow` for it to work, and I just felt it was too much of a change.
Also, consider that in order for this to work, the visitors need to be `nothrow` / whatever attributes we fancy, which greatly reduce the usability of DMD as a library.

Comment (Geod24) [reply src/dmd/semantic2.d]: Note that this should *never* be triggered. Indeed DMD does not `throw`. `scope (failure) assert(0);` is a cheap trick to make the function `nothrow`.

**Comment by Geod24** [comment]
Okay that's not going to work. Maybe it's time to have another CI provider run the C++ interop tests.

**Comment by MoonlightSentinel** [comment]
> Okay that's not going to work. Maybe it's time to have another CI provider run the C++ interop tests.

Or manually let the test suite fetch a more recent clang if necessary?

File: test/runnable/cpp11.d [hunk 3]
```diff
@@ -61,10 +61,66 @@ void test19658()
...
+// but the autotest only has 3.5.
+version (OSX)
+{
+    private struct ABITags
+    {
+        this(string[] tags...) {}
+    }
+}
+else
+    alias ABITags = gnuAbiTag;
```
Comment (MoonlightSentinel) [inline test/runnable/cpp11.d]: This test is only disabled on win32, hence the Azure failure

**Comment by Geod24** [comment]
> Okay that's not going to work. Maybe it's time to have another CI provider run the C++ interop tests.

https://github.com/dlang/dmd/pull/10964

File: src/dmd/semantic2.d [hunk 4]
```diff
@@ -558,14 +569,18 @@ private extern(C++) final class Semantic2Visitor : Visitor
                         if (e.op == TOK.tuple)
                         {
                             TupleExp te = cast(TupleExp)e;
-                            eval(sc, te.exps);
+                            eval(sc, te.exps, lastTag);
                         }
+
+                        // Handles compiler-recognized `core.attribute.gnuAbiTag`
+                        if (UserAttributeDeclaration.isGNUABITag(e))
+                            doGNUABITagSemantic(e, lastTag);
```
Comment (ibuclaw) [inline src/dmd/semantic2.d]: The condition could be moved to either Compiler or Target (probably the former is best) so other compilers can extend their own compiler recognised udas - if I understand that this is the correct place to do so.

Comment (ibuclaw) [reply src/dmd/semantic2.d]: Ah wait, I didn't get to the bit where it's a library defined uda.

In which case, there should be tests that ensure bad things don't happen if people supply their own (incorrect) type.

File: posix.mak [hunk 5]
```diff
@@ -19,7 +19,14 @@ ifneq (,$(findstring Darwin_64_32, $(PWD)))
 auto-tester-test:
 	echo "Darwin_64_32_disabled"
 else
+ifneq (,$(findstring windows, $(PWD)))
 auto-tester-test: test
+else # POSIX
+# Like test, but without runnable_cxx
+auto-tester-test:
+	$(QUIET)$(MAKE) -C src -f posix.mak unittest
+	$(QUIET)$(MAKE) -C test -f Makefile
```
Comment (MoonlightSentinel) [inline posix.mak]:
```suggestion
	$(QUIET)$(MAKE) -C test -f Makefile auto-tester-test
```

Comment (Geod24) [reply posix.mak]: Thanks 🤦‍♂

File: test/compilable/cpp_abi_tag_unused.d [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/* DISABLED: win32 win64
+REQUIRED_ARGS: -extern-std=c++11
+TEST_OUTPUT:
+---
+---
```
Comment (MoonlightSentinel) [inline test/compilable/cpp_abi_tag_unused.d]: An empty `TEST_OUTPUT` section isn't required anymore.

File: test/tools/d_do_test.d [hunk 7]
```diff
@@ -1550,3 +1557,26 @@ void printTestFailure(string testLogName, string output_file_temp)
...
+ *
+ * If linking failed, we print the symbols present in the C++ object file being
+ * linked it. This is so that C++ `runnable` tests are easier to debug,
+ * as the CI machines can have different environment than the users,
+ * and it is generally painful to work with them when trying to support
+ * newer (C++11, C++14, C++17, etc...) features.
+ */
+void printCppSources (in const(char)[][] compiled)
+{
+    version (Posix)
```
Comment (MoonlightSentinel) [inline test/tools/d_do_test.d]: The windows machines have some ported Posix utilities AFAICT (e.g. bash). Might be worth to try `nm` and silently ignore failures on windows.

**Comment by Geod24** [comment]
Runnable_cxx also needs to be disabled on semaphore ci because it uses GCC 4.8.4

**Comment by Geod24** [comment]
Green at last 🎉

**Review by thewilsonator — APPROVED** [review]

**Comment by Geod24** [comment]
Hum, this went red for Ubuntu after merge... Will fix.

**Comment by Geod24** [comment]
Fix being worked on here: https://github.com/dlang/dmd/pull/10974

**Pull request merged**